test/cmd/lfstest-customadapter.go

```diff
@@ -0,0 +1,253 @@
+// +build testtools
+
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/github/git-lfs/api"
+	"github.com/github/git-lfs/httputil"
+	"github.com/github/git-lfs/progress"
+	"github.com/github/git-lfs/tools"
+)
+
+// This test custom adapter just acts as a bridge for uploads/downloads
+// in order to demonstrate & test the custom transfer adapter protocols
+// All we actually do is relay the requests back to the normal storage URLs
+// of our test server for simplicity, but this proves the principle
+func main() {
+
+	scanner := bufio.NewScanner(os.Stdin)
```

Comment:
Robustness, since we're dealing with cross-process comms and need to not assume how the other end flushes the streams. For example if the sender wrote 2 messages in quick succession to the stream and called flush only after both, a plain Read() could read both at once and you need somewhere to stash the second - to cope with this would basically be doing exactly what `bufio` code does.